File: src/main/java/net/rayfall/eyesniper2/skRayFall/EffectLibSupport/EffEffectLibArc.java
package net.rayfall.eyesniper2.skRayFall.EffectLibSupport;

import net.rayfall.eyesniper2.skRayFall.skRayFall;

import org.bukkit.Location;
import org.bukkit.entity.Entity;
import org.bukkit.event.Event;
import org.eclipse.jdt.annotation.Nullable;

import ch.njol.skript.lang.Effect;
import ch.njol.skript.lang.Expression;
import ch.njol.skript.lang.SkriptParser.ParseResult;
import ch.njol.util.Kleenean;
import de.slikey.effectlib.effect.ArcEffect;
import de.slikey.effectlib.util.DynamicLocation;
import de.slikey.effectlib.util.ParticleEffect;

public class EffEffectLibArc extends Effect{
	
	//(spawn|create|apply) (a|the|an) arc (effect|formation) from %entity/location% to %entity/location% with id %string% [with particle[s] %-effectlibparticle%] [%number% high] [and %number% particles]

	private Expression<?> start;
	private Expression<?> end;
	private Expression<String> id;
	private Expression<ParticleEffect> particle;
	private Expression<Number> height;
	private Expression<Number> numOfParticles;
	
	@SuppressWarnings("unchecked")
	@Override
	public boolean init(Expression<?>[] e, int arg1, Kleenean arg2,
			ParseResult arg3) {
		start = e[0];
		end = e[1];
		id = (Expression<String>) e[2];
		particle = (Expression<ParticleEffect>) e[3];
		height = (Expression<Number>) e[4];
		numOfParticles = (Expression<Number>) e[5];
		return true;
	}

	@Override
	public String toString(@Nullable Event arg0, boolean arg1) {
		return null;
	}

	@Override
	protected void execute(Event evt) {
		Object s = start.getSingle(evt);
		Object en = end.getSingle(evt);
		ArcEffect effect = new ArcEffect(skRayFall.effectManager);
		if (s instanceof Entity) {
			effect.setDynamicOrigin(new DynamicLocation((Entity) s));
		} else if (s instanceof Location) {
			effect.setDynamicOrigin(new DynamicLocation((Location) s));
		} else {
			assert false;
		}
		if (en instanceof Entity) {
			effect.setDynamicTarget(new DynamicLocation((Entity) en));
		} else if (en instanceof Location) {
			effect.setDynamicTarget(new DynamicLocation((Location) en));
		} else {
			assert false;
		}
		if(particle != null){
			effect.particle = particle.getSingle(evt);
		}
		if(height.getSingle(evt) != null){
			effect.height = height.getSingle(evt).floatValue();
		}
		if(numOfParticles.getSingle(evt) != null){
			effect.particles = numOfParticles.getSingle(evt).intValue();
		}
		effect.infinite();
		effect.start();
		Boolean check = skRayFall.effLibManager.setEffect(effect, id.getSingle(evt).replace("\"", ""));
		if (!check) {
			effect.cancel();
		}
	}

}

File: src/main/java/net/rayfall/eyesniper2/skrayfall/effectlibsupport/EffEffectLibArc.kt
package net.rayfall.eyesniper2.skrayfall.effectlibsupport

import ch.njol.skript.Skript
import ch.njol.skript.doc.Description
import ch.njol.skript.doc.Name
import ch.njol.skript.lang.Effect
import ch.njol.skript.lang.Expression
import ch.njol.skript.lang.SkriptParser
import ch.njol.skript.util.VisualEffect
import ch.njol.util.Kleenean
import de.slikey.effectlib.effect.ArcEffect
import de.slikey.effectlib.util.DynamicLocation
import net.rayfall.eyesniper2.skrayfall.Core
import org.bukkit.Location
import org.bukkit.entity.Entity
import org.bukkit.event.Event

@Name("Arc Effect")
@Description("Creates an EffectLib arc effect.")
class EffEffectLibArc : Effect() {

    // (spawn|create|apply) (a|the|an) arc (effect|formation) from %entity/location% to
    // %entity/location% with id %string% [with particle[s] %-effectlibparticle%] [%number% high] [and
    // %number% particles]

    private var startExpression: Expression<*>? = null
    private var endExpression: Expression<*>? = null
    private var idExpression: Expression<String>? = null
    private var particleExpression: Expression<VisualEffect>? = null
    private var heightExpression: Expression<Number?>? = null
    private var numberOfParticlesExpression: Expression<Number?>? = null

    @Suppress("UNCHECKED_CAST")
    override fun init(exp: Array<Expression<*>?>, arg1: Int, arg2: Kleenean, arg3: SkriptParser.ParseResult): Boolean {
        startExpression = exp[0]
        endExpression = exp[1]
        idExpression = exp[2] as Expression<String>
        particleExpression = exp[3] as? Expression<VisualEffect>
        heightExpression = exp[4] as Expression<Number?>
        numberOfParticlesExpression = exp[5] as Expression<Number?>
        return true
    }

    override fun toString(arg0: Event?, arg1: Boolean): String {
        return ""
    }

    override fun execute(evt: Event) {
        val startObject = startExpression?.getSingle(evt)
        val endObject = endExpression?.getSingle(evt)
        val particle = EffectLibUtils.getParticleFromVisualEffect(particleExpression?.getSingle(evt))
        val height = heightExpression?.getSingle(evt)
        val numberOfParticles = numberOfParticlesExpression?.getSingle(evt)
        val id = idExpression?.getSingle(evt)
        val baseEffect = ArcEffect(Core.effectManager)

        if (id == null)
        {
            Skript.warning("Id was null for EffectLib Arc")
            return
        }

        when (startObject) {
            is Entity -> {
                baseEffect.setDynamicOrigin(DynamicLocation(startObject as Entity?))
            }
            is Location -> {
                baseEffect.setDynamicOrigin(DynamicLocation(startObject as Location?))
            }
            else -> {
                assert(false)
            }
        }

        when (endObject) {
            is Entity -> {
                baseEffect.setDynamicTarget(DynamicLocation(endObject as Entity?))
            }
            is Location -> {
                baseEffect.setDynamicTarget(DynamicLocation(endObject as Location?))
            }
            else -> {
                assert(false)
            }
        }

        if (particle != null) {
            baseEffect.particle = particle
        }
        if (height != null) {
            baseEffect.height = height.toFloat()
        }
        if (numberOfParticles != null) {
            baseEffect.particles = numberOfParticles.toInt()
        }

        baseEffect.infinite()
        baseEffect.start()

        val setEffectSuccess = Core.rayfallEffectManager.setEffect(baseEffect, id.replace("\"", ""))
        if (!setEffectSuccess) {
            baseEffect.cancel()
        }
    }
}